### PR TITLE
Add support for materialized CTEs

### PR DESCRIPTION
**Pull request by maartenvanvliet**
Adds support for MATERIALIZED CTE's. Before Postgres 12 CTE's were materialized by default but from then on it was opt-in  with the MATERIALIZED keyword. In some cases this can be beneficial because then the CTE can act as a optimisation fence, which can result in better query plans sometimes.

This PR still needs some work, so I'm opening it to see if there's interest in this. If so, have I taken the right approach here?

Corresponding echo_sql PR https://github.com/elixir-ecto/ecto_sql/pull/466

See also https://www.postgresql.org/docs/current/queries-with.html#id-1.5.6.12.7 and https://paquier.xyz/postgresql-2/postgres-12-with-materialize/

### CODE REVIEW

File: lib/ecto/query.ex
Comment (maartenvanvliet) [inline lib/ecto/query.ex:1239]: The `as` keyword now needs to be the first list item. Is that desired?

Comment (v0idpwn) [reply lib/ecto/query.ex:1239]: Based in the docs from https://www.postgresql.org/docs/15/queries-with.html (section 7.8) the behaviour is a bit more complex than it may seem at first:
- Not specifying either `MATERIALIZED` or `NOT MATERIALIZED` will leave up to postgres whether to materialize
- Specifying overrides the postgres selection, which can bring performance improvements.
So we would need to support both options.

Comment (maartenvanvliet) [reply lib/ecto/query.ex:1239]: Shall we code it with `:not_materialized | :materialized | nil` internally? `nil` would leave it up to postgres whether to materialize.

Comment (v0idpwn) [reply lib/ecto/query.ex:1239]: I'm not sure, cc: @josevalim, @greg-rychlewski

Comment (greg-rychlewski) [reply lib/ecto/query.ex:1239]: Can't it be like this

```elixir
with_cte(name, query, opts)

## options
  `:as` - the CTE query itself or a fragment
  `:materialize - a boolean indicating whether the CTE should be materialized. if blank, the database's default behaviour will be used
```  

Also I don't think there's a need to force the keyword list order. You can just make sure they specify `:as` and raise if not.

Comment (josevalim) [reply lib/ecto/query.ex:1239]: :+1: for @greg-rychlewski's proposal!

File: lib/ecto/query/builder/cte.ex
Comment (greg-rychlewski) [inline lib/ecto/query/builder/cte.ex:40]:
```suggestion
  def build(query, name, cte, materialized?, env) do
    Builder.apply_query(query, __MODULE__, [escape(name, env), build_cte(name, cte, env), materialized?], env)
```

I think the convention is to add a trailing question mark  but not 100% sure. I'm thinking of things like `changeset.valid?`. cc @josevalim

**Review by greg-rychlewski — COMMENTED** [review]
Something is a bit off, some of the tests that should be there are not in your branch. Like this one in the planner test:

```elixir
   test "with CTEs" do
      cte_query = from(s in "schema", select: %{x1: selected_as(s.x, :integer), x2: s.x})

      %{with_ctes: %{queries: [{"schema_cte", inner_query}]}} =
        Comment
        |> with_cte("schema_cte", as: ^cte_query)
        |> normalize()

      assert [{:integer, _}, {:x2, _}] = inner_query.select.fields
    end
```
By any chance is your fork out of date? would you be able to rebase it?

File: test/ecto/query/builder/cte_test.exs
Comment (greg-rychlewski) [inline test/ecto/query/builder/cte_test.exs:20]: we should add a test for materialized = false as well once that change has been made

**Review by greg-rychlewski — APPROVED** [review]
Looks good to me thank you. I'll just let @v0idpwn or @josevalim approve as well. Then once this is merged we can point your ecto_sql PR back to it.

File: lib/ecto/query.ex
Comment (v0idpwn) [inline lib/ecto/query.ex:1221]: Not sure, but could be worth mentioning which adapters support this feature.

File: lib/ecto/query.ex
Comment (v0idpwn) [inline lib/ecto/query.ex:1298]:
```suggestion
      Builder.error! "`:as` option must be specified"
```

**Review by v0idpwn — COMMENTED** [review]
LGTM.

File: lib/ecto/query.ex
Comment (v0idpwn) [reply lib/ecto/query.ex:1298]: With this fixed its approved from my side :)

**Comment by maartenvanvliet** [comment]
Thx everyone! Happy holidays! 🎅

**Pull request merged**
